### DISCUSSION TITLE
[Proposal]: Add a narrower than internal access modifier for nested class' members

### DISCUSSION THREAD

**Discussion by martinrhan** [dotnet/csharplang#5554] [Language Ideas]
# FEATURE_NAME

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
[summary]: #summary
When we define class A inside class B, the outer class B cannot access class A field if its private. However, sometimes making it internal is to much because we don't need them to be visible to the whole assembly.

## Motivation
[motivation]: #motivation
This proposal can improve code quality and make the language have better capability of encapsulation. 
Lets say we have a class A which has some field that should only be visible limited in a certain scope.
```
public class A { // This class should be visible for everyone
    public field a; // There are several public members
    internal field b; // Members that have limited access.
}
```
In the above example, member b is visible in the current assembly but not at outside. Similarly, inside an assembly, we may have some modules that perform particular function and should be decoupled from other part of the assembly.
```
internal class A{ //This class should be visible for everyone in the assembly
    internal field a; // There are several members visible for everyone in the assembly
    ??? field b; // Members that have limited access, cannot be private, but internal is too much.
}
``` 
In other languages, a method to attain that is friend class. However, seems like its depreciated in C#. I saw people recommend nested class, in which outer class' private members can be accessed by inner class. This is good but not enough. Because sometimes the inner class need to be the one depended by outer class and thus outer class need to visit inner class member. 
```
internal class B{
    internal class A{
        private field a; // not accessible by B
        internal field b // accessible by B but also accessible by other classes in this assembly.
        ??? field c // accessible by B but not accessible by other classes in this assembly, which is what we need.
    }
}
```
As shown above, we need the members allowed to be somewhere in bettween, that is, accessible by outer class and current class only.

## Detailed design
[design]: #detailed-design
There is not to much to talk about how to design, just add another keyword, maybe private internal or something else.

## Drawbacks
[drawbacks]: #drawbacks
I cannot figure out any drawbacks, this should be purely beneficial.

## Alternatives
[alternatives]: #alternatives
As mensioned already, an alternative can be friend class. Further more, by existing language feature, we can attain that by exposing public  (or internal) abstract class and implement it with a private class:
```
internal class B{
    internal void MethodB(){
         AImplementation A = new();
         A.a = something;
    }
    internal abstract class ABase{
        internal abstract void MethodA(); //Accessible for the whole assembly
    }
   private class AImplementation : ABase{
        internal field a; //Accessible in class B and this class only
        internal override void Method A(){} //Accessible for the whole assembly
   }
}
```
However, this method is a trick that adds many unecessary code which could be simply done by a simple access modifier. More importantly, this method doesn't work if the nested type is a struct, which doesn't have inheritance.

## Unresolved questions
[unresolved]: #unresolved-questions

**Comment by 333fred** [comment]
As detailed in the README, we ask that new topics be opened first as discussions, which are better suited to discussing new ideas that are not currently championed by a member of the LDM.

**Comment by theunrepentantgeek** [comment]
It's unclear to me what you're asking for.

If you want the enclosing class to have access to members without allowing external consumers to access the fields, just [make the nested class **private**](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABATARgLABQGAzAATakDCpA3oaQ+WRiqQLIAUAlLfY/wDcAhlFIA7UgF5xMAO6kAcjADOAFxgATbgG4+/BmIB0kAK5jVU0iiy6C/AL57STgA5QAlsPXksileo0nOjt9BhJSd3NSU3NbB0J7IA=):

``` csharp
public class C {
    public void M() {
        var n = new Nested();
        n.count = 42;
    }
    
    private class Nested
    {
        public int count;
    }
}
```

If you want to expose an inner type (why?) then you can make an **public** **abstract** base class to return, and then [create a private implementation class](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABATARgLABQGAzAATakDCpA3oaQ+WQHIwDOALjACakCyACgCUteo3EA3AIZRSAO1IBeeTADupVpx4BJALYAHADbCA3GPEM5AOkgBXORyWkUWMwQuMMAdnlvxAX3NSIJJSKWBOKCkwRwpNLl4guncPJjCIjiiY0gBLB1IAMQgIUyDAlIYg/Sgc6S5yLA12BL0jECatbiSg8VC8xzsHNx7PMggJGCga7hhc/KKSkUUAPlJBjj9Gcv8gA)

``` csharp
using System;
public class C {
    public Nested M() {
        var n = new NestedImpl();
        n.count = 42;
        return n;
    }
    
    public abstract class Nested 
    {
        public abstract int Foo();
    }
    
    private class NestedImpl: Nested
    {
        public int count;

        public override int Foo() => count;
    }
}
```

Do either of these achieve what you're wanting? If not, perhaps you'd clarify what behaviour you're wanting?

**Reply by martinrhan** [reply]
your second example is the one corresponding to what I am arguing. Its is good, but still not enought. First of all, constructor is not available for public userers. This doesn't matter too much as we can implement a factory method.
Another reason is important, this cannot be done if the inner type is a struct. Struct doesn't have enheritance. My code have structs nested.

**Reply by martinrhan** [reply]
I updated the post to make it more clear

**Comment by fa1ntl1ght** [comment]
I would like full friend classes support instead

**Comment by airbreather** [comment]
I just came across a use case for this yesterday, so I figured I'd jot it down here.

For various reasons, I've come across a situation where I want to expose something like* a `Span<MyValueType>`, where the actual contents of `Span<MyValueType>` are to be treated as nearly opaque except for a single `public ulong ItemId;` field.  The trick here is that I've got it set up in such a way where `MyValueType` is defined as a struct that's nested in the static class that provides all the high-level methods to manipulate the contents of the `Span<MyValueType>`.

**It's not **exactly** `Span<MyValueType>`, but rather something I've rolled myself to achieve similar goals.  For all intents and purposes, let's call it that, but do note that it's not stack-only.*

A bit more concrete:
```csharp
public static class TreeHelpers
{
    public struct Node
    {
        public ulong ItemId;

        internal double ImplementationValue1;

        internal ulong ImplementationValue2;

        internal ulong ImplementationValue3;
    }

    public static void BuildTreeStructure(Span<Node> nodes)
    {
        // caller must set the ItemId on each node before calling this
        // we will reorder the values in the span and also set the implementation values.
    }

    public static double VisitTree(ReadOnlySpan<Node> nodes, SomeVisitor visitor)
    {
        // caller gives us the nodes that they built before, potentially having been saved / loaded
        // we use the tree structure and ImplementationValue1/2/3 to send the ItemId values over to the visitor as we walk the tree.
    }
}
```
I kinda like the idea of nesting the `Node` struct inside the static class that contains all the methods that make it actually useful.  Nobody outside of the `TreeHelpers` static class has any business engaging with any specific field except `ItemId`, but because all the callers **happen** to be in the same assembly, they can all see them.  *In fact, the actual `TreeHelpers` static class itself is `internal`, because it's an implementation detail of things that provide functionality that's at an even higher level than this.*

I will be the first to admit that the design for this specific use case is not idiomatic.  This particular application as a whole tends to see a lot of non-idiomatic patterns, due to several extreme constraints that I think are out-of-scope for this discussion.  However, I think that the general concept here isn't exactly all that far-out (it's something I recall doing before in more idiomatic applications, I just don't have a fresh example):

I want to be able to create data structures, nested within a type that encapsulates all the methods that make them useful, where **some** of that data structure's members can be seen and manipulated outside of that higher-level type, and **others** can only be seen and manipulated by the implementations of those methods.

I see that #2073 could also be used to achieve these goals, but I do slightly lean towards agreeing with the general sentiment on that thread that nesting is a more natural way to define the relationship that I'm looking for here, despite not being a full solution to all the problems that people cite over there.
